org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/jdt/JavaDebugServer.java

```diff
@@ -23,90 +23,118 @@
+    private static class InternalDebugServer {
+        private ServerSocket serverSocket = null;
+        private boolean isStarted = false;
+        private String serverName;
+
+        InternalDebugServer(String serverName) {
             try {
-                serverSocket.close();
+                this.serverName = serverName;
+                this.serverSocket = new ServerSocket(0, 1);
             } catch (IOException e) {
-                Logger.logException("Close ServerSocket exception", e);
+                Logger.logException("Failed to create Internal Debug Server", e);
             }
         }
-        serverSocket = null;
-    }
 
-    private void closeConnection() {
-        if (connection != null) {
-            try {
-                connection.close();
-            } catch (IOException e) {
-                Logger.logException("Close client socket exception", e);
+        synchronized int getPort() {
+            if (this.serverSocket != null) {
+                return this.serverSocket.getLocalPort();
             }
+            return -1;
+        }
+
+        synchronized void start() {
+            if (this.serverSocket != null && !this.isStarted) {
+                this.isStarted = true;
+                // Execute eventLoop in a new thread.
+                new Thread(new Runnable() {
+
+                    @Override
+                    public void run() {
+                        while (true) {
+                            try {
+                                // Allow server socket to service multiple clients at the same time.
+                                // When a request comes in, create a connection thread to process it.
+                                // Then the server goes back to listen for new connection request.
+                                Socket connection = serverSocket.accept();
+                                createConnectionThread(connection);
+                            } catch (IOException e1) {
+                                Logger.logException("Setup socket connection exception", e1);
+                                closeServerSocket();
+                                return;
+                            }
+                        }
+                    }
+
+                }, this.serverName).start();
+            }
+        }
+
+        synchronized void stop() {
+            closeServerSocket();
+        }
+
+        private void closeServerSocket() {
+            if (serverSocket != null) {
+                try {
+                    Logger.logInfo("Close debugserver socket port " + serverSocket.getLocalPort());
+                    serverSocket.close();
+                } catch (IOException e) {
+                    Logger.logException("Close ServerSocket exception", e);
+                }
+            }
+            serverSocket = null;
+        }
+
+        private void createConnectionThread(Socket connection) {
+            new Thread(() -> {
```

Comment:
The thread will have the same life cycle as the socket connection. And keep alive when the Socket connection is alive. Die when socket connection is closed. So the thread is managed by socket connection potentially.